Levenshtein & Hamming Distance

c5sire opened the pull request:
This adds the Levenshtein function. It uses an algorithm that does not use an internal array and should be faster and more memory efficient. For the time being, string length is limited to 254. 
Addresses issue #1435.

Also, a simple mismatch counting function was added addressing issue #1439.

simonasked committed:
Implemented mismatches function

simonasked committed:
adding levenshtein

simonasked committed:
Added Levenshtein function and tests

c5sire commented:
I am trying to use the automated format fix script, but apparently it is not working on my fork. Any advice?
Best,
Reinhard

? make format-fix
python3 scripts/format.py --all --fix
Formatting all files
The files listed above will be reformatted.
Continue with changes (y/n)?
y
sh: 1: cmake-format: not found
Traceback (most recent call last):
  File "scripts/format.py", line 270, in <module>
    format_directory('src')
  File "scripts/format.py", line 265, in format_directory
    f.split('.')[-1])
  File "scripts/format.py", line 227, in format_file
    new_text = get_formatted_text(f, full_path, directory, ext)
  File "scripts/format.py", line 210, in get_formatted_text
    proc = subprocess.Popen(proc_command, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
  File "/mnt/d/apps/ubuntu/bin/miniconda3/lib/python3.7/subprocess.py", line 800, in __init__
    restore_signals, start_new_session)
  File "/mnt/d/apps/ubuntu/bin/miniconda3/lib/python3.7/subprocess.py", line 1551, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'cmake-format': 'cmake-format'
make: *** [Makefile:157: format-fix] Error 1
? make format
make: Nothing to be done for 'format'.

Mytherin commented:
You need to install cmake-format, try `pip install cmake-format`. We should probably skip formatting cmake files if cmake-format is not installed instead of failing.

Mytherin commented on src/function/scalar/string/levenshtein.cpp:
```diff
@@ -0,0 +1,84 @@
+
+    std::vector<int> distances(txt.size() + 1, 0);
+
+    for (int pos = 1; pos <= (int)txt.size(); pos++) 
```
Could you change the indexes used in the different loops to `idx_t` rather than `int`? We prefer to avoid `int` since its size is platform-dependent.

c5sire replied:
Certainly.

Mytherin commented on src/function/scalar/string/levenshtein.cpp:
```diff
@@ -0,0 +1,84 @@
+    {
+        return -1; // too short
+    }
+	if (txt.size() > 254 || tgt.size() > 254) 
```
What exactly is the reason for the 255 byte limitation? It seems like the code below works fine for larger strings, no?

c5sire replied:
No specific reason. Just thought it might be cautious. I will remove it.

Mytherin commented on src/function/scalar/string/levenshtein.cpp:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    int dist = 0;
+	int tempDist = 0;
```
idx_t here as well please :)

Mytherin commented on src/function/scalar/string/levenshtein.cpp:
```diff
@@ -0,0 +1,84 @@
+// See: https://www.kdnuggets.com/2020/10/optimizing-levenshtein-distance-measuring-text-similarity.html
+// And: Iterative 2-row algorithm: https://en.wikipedia.org/wiki/Levenshtein_distance
+// Note: A first implementation using the array algorithm version resulted in an error raised by duckdb (too muach memory usage) 
+static int levenshtein2(const string& txt, const string& tgt) 
```
We prefer to avoid the use of `std::string` since it performs a large number of (unnecessary) allocations. Could you change this function to operate on our internal `string_t` type instead? The resulting code would look very similar, only using `GetSize()` instead of `size()`, and `GetDataUnsafe()` to obtain a raw pointer to the data (note that string_t is not null-terminated).

c5sire replied:
Thanks for the pointers.

Mytherin commented on src/function/scalar/string/mismatches.cpp:
```diff
@@ -0,0 +1,47 @@
+	int64_t mismatches = 0;
+
+	for (int64_t idx = 0; idx < (int64_t)str.GetSize(); ++idx) {
+		if (str.GetString()[idx] != tgt.GetString()[idx]) {
```
Please avoid using `GetString()`, especially in a loop such as this: two new strings will be constructed for every character in the string. Instead, use `GetDataUnsafe()` and store the resulting pointer in a temporary variable, e.g.:


```cpp
auto str_data = str.GetDataUnsafe();
auto tgt_data = tgt.GetDataUnsafe();
for (...) {
    if (str_data[idx] != tgt_data[idx]) { ... }
}
```

c5sire replied:
Thanks!

Mytherin commented on src/function/scalar/string/mismatches.cpp:
```diff
@@ -0,0 +1,47 @@
+	int64_t num_characters_str = LengthFun::Length<string_t, int64_t>(str);
+	int64_t num_characters_tgt = LengthFun::Length<string_t, int64_t>(tgt);
+	if (num_characters_str != num_characters_tgt) {
+		return -1;  // Expected error code: both strings must be of equal length
```
If you wish to throw an error from a function you should throw an exception, e.g. `throw InvalidInputException("...")`. This will report the error to the user.

c5sire replied:
ok

Mytherin commented on test/sql/function/string/test_levenshtein.test:
```diff
@@ -0,0 +1,22 @@
+
+# normal length
+query I
+SELECT levenshtein(s, 'hallo') FROM strings ORDER BY s
```
I would like to see a number of extra tests, including `NULL` values, and each of the different combinations of (CONSTANT, CONSTANT), (COLUMN, CONSTANT), (CONSTANT, COLUMN), (COLUMN, COLUMN).

In addition, I would like to see tests that trigger each of the possible error conditions (too short strings, empty strings, too long strings).

c5sire replied:
Noted.

Mytherin commented on test/sql/function/string/test_mismatches.test:
```diff
@@ -0,0 +1,24 @@
+
+# normal length
+query I
+SELECT mismatches(s, 'hallo') FROM strings ORDER BY s
```
Same here, please add more tests as described above

Mytherin reviewed (commented):
Thanks for your PR! This is looking very good! I have the following comments on the PR. When you have time could you have a look at them? Cheers.

simonasked committed:
refactoring

simonasked committed:
Adding more tests for mismatches function

simonasked committed:
revising test

simonasked committed:
refactoring levenshtein

simonasked committed:
adding levenshtein tests

simonasked committed:
Addressed code review comments
Refactored Levenshtein algorithm
Added more test cases
Added alias for 'levenshtein': editdist3 (as in sqlite)
Added alias for 'mismatches': hamming

simonasked committed:
Tidied code

simonasked committed:
Tidy and format fixes

simonasked committed:
more formatting

simonasked committed:
format fixes

simonasked committed:
more fixes

simonasked committed:
Tidied mismatch

simonasked committed:
formatting ctd

simonasked committed:
adding more tests

simonasked committed:
adding tests for two columns

simonasked committed:
Fixing tests

simonasked committed:
format fixes

c5sire commented:
Addressed your review comments.
I noticed that after fixing formatting and tidying issues, there are now apparently inconsistent failures with some environments like windows32, windows64, and Raspberry Pi. I didn't see any direct connection with the code I was working on, though.

hannesmuehleisen commented:
Indeed the only failing test that remains is the Windows 64 bit test. I see
````
D:/a/duckdb/duckdb/test/api/test_progress_bar.cpp(36): FAILED:
117
  REQUIRE( con.context->progress_bar->IsPercentageValid() )
118
with expansion:
119
````

Maybe @pdet can have a look?

c5sire commented:
Similarly a test failed for PR#1451 for windows32:
D:/a/duckdb/duckdb/test/api/test_progress_bar.cpp(36): FAILED:
  REQUIRE( con.context->progress_bar->IsPercentageValid() )
with expansion:
  false

pdet commented:
Hi, yes, I can confirm that the progress bar is non-deterministically breaking on windows, I'm looking into it!

Mytherin commented on .clang-format:
```diff
@@ -15,7 +15,6 @@ SpaceBeforeCtorInitializerColon: true
 SpaceBeforeInheritanceColon: true
 SpacesInAngles: false
 SpacesInCStyleCastParentheses: false
-SpacesInConditionalStatement: false
```
Was this included by accident, or did you intend to change the formatting rules here? If so, could you expand on why?

c5sire replied:
It caused an error on my local installation and the script would not proceed.

Mytherin replied:
Perhaps you are using an older version of clang-format? We are using version 11 (`clang-format version 11.0.1`).

c5sire replied:
Re-added the line.

Mytherin commented on src/function/scalar/string/levenshtein.cpp:
```diff
@@ -0,0 +1,98 @@
+namespace duckdb {
+
+static idx_t Minimum(const idx_t &x, const idx_t &y, const idx_t &z) {
+	return std::min(x, std::min(y, z));
```
Please avoid using generic names like `Minimum` for functions. In DuckDB we support amalgamation builds which means all source files will be combined into a single `.cpp` file. Generic names are a potentially source of conflicts, since `Minimum` might be used somewhere else in the codebase. 

Since the function is only used in one place, I recommend inlining the function.

Also please use `MinValue` instead of std::min (from [duckdb/common/helper.hpp](https://github.com/cwida/duckdb/blob/master/src/include/duckdb/common/helper.hpp)). `std::min` has bitten us before on certain platforms.

c5sire replied:
Noted.

c5sire replied:
Used MinValue and simplified.

Mytherin reviewed (commented):
Thanks for the changes again, this is looking much better now. Some minor comments, and then I think this is ready to be merged:

simonasked committed:
Refactored minimum

simonasked committed:
Re-adding format rule

Mytherin commented:
Thanks for all the changes, looks great now!

c5sire commented:
Thanks to you!

c5sire commented:
Hi, 
I noticed that my levenshtein, hamming, and jaccard functions do not work in the WHERE clause of a statement. While other text functions like length do. What did I miss? 
Best
